### PR TITLE
Index functions

### PR DESCRIPTION
**Pull request by menakaj**

### CODE REVIEW

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexConstructorUtil.java [hunk 1]
```diff
@@ -175,18 +171,15 @@ public static boolean readableXmlFile(String path) {
...
-                                                       ITreeNodeIdProvider nodeIdProvider,
-                                                       ByteBufferInputStream bbis, DataInputStream di, String nodeId)
-            throws IOException {
+    public static IndexDocumentBuilder getIndexBuilder(File file, IndexWriter writer, TaggedValuePointable nodep,
+            ArrayBackedValueStorage abvsFileNode, ITreeNodeIdProvider nodeIdProvider, ByteBufferInputStream bbis,
+            DataInputStream di, String nodeId) throws IOException {
 
         //Get the document node
-        IParser parser = new XMLParser(false, nodeIdProvider, nodeId);
+        XMLParser parser = new XMLParser(false, nodeIdProvider, nodeId);
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexConstructorUtil.java]: Why change from IParser?

File: vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-functions.xml [hunk 2]
```diff
@@ -139,16 +139,24 @@
     <!-- fn:update-index($indexFolder  as xs:string?) as  node()* -->
     <function name="fn:update-index">
         <param name="index-folder" type="xs:string?"/>
-        <return type="node()*"/>
+        <return type="xs:boolean?"/>
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-functions.xml]: xs:boolean (don't make this optional)

File: vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-functions.xml [hunk 3]
```diff
@@ -139,16 +139,24 @@
...
         <runtime type="scalar"
                  class="org.apache.vxquery.runtime.functions.index.IndexUpdaterEvaluatorFactory"/>
     </function>
 
+    <!-- fn:delete-index($indexFolder  as xs:string?) as  node()* -->
+    <function name="fn:delete-index">
+        <param name="index-folder" type="xs:string?"/>
+        <return type="xs:boolean?"/>
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-functions.xml]: xs:boolean (don't make this optional)

File: vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-functions.xml [hunk 4]
```diff
@@ -139,16 +139,24 @@
...
+        <runtime type="scalar"
+                 class="org.apache.vxquery.runtime.functions.index.IndexDeleteEvaluatorFactory"/>
+    </function>
+
     <!-- fn:collection-from-index($indexfolder  as xs:string?, $elementpath as xs:string?) as  node()* -->
     <function name="fn:collection-from-index">
         <param name="index-folder" type="xs:string?"/>
         <param name="element-path" type="xs:string?"/>
-        <return type="node()*"/>
+        <return type="xs:boolean?"/>
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-functions.xml]: xs:boolean (don't make this optional)

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexConstructorScalarEvaluatorFactory.java [hunk 5]
```diff
@@ -61,8 +64,12 @@ protected IScalarEvaluator createEvaluator(IHyracksTaskContext ctx, IScalarEvalu
...
             @Override
             protected void evaluate(TaggedValuePointable[] args, IPointable result) throws SystemException {
-                IndexConstructorUtil.evaluate(args, result, stringp, bbis, di, sb, abvs, nodeIdProvider, abvsFileNode,
-                        nodep, false, nodeId);
+                try {
+                    IndexConstructorUtil.evaluate(args, result, stringp, bbis, di, sb, abvs, nodeIdProvider, abvsFileNode,
+                            nodep, false, nodeId);
+                } catch (JAXBException e) {
+                    throw new SystemException(ErrorCode.SYSE0001, e);
+                }
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexConstructorScalarEvaluatorFactory.java:74]: Add the boolean result:

XDMConstants.setTrue(result);

Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexConstructorScalarEvaluatorFactory.java:74]: Also modify the other methods for update and delete.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexUpdaterEvaluatorFactory.java [hunk 6]
```diff
@@ -65,9 +68,12 @@ protected void evaluate(TaggedValuePointable[] args, IPointable result) throws S
...
-                    updater.evaluate();
-                } catch (IOException | NoSuchAlgorithmException e) {
-                    e.printStackTrace();
+                    updater.setup();
+                    updater.updateIndex();
+                    updater.updateMetadataFile();
+                    updater.exit();
+                } catch (IOException | NoSuchAlgorithmException | JAXBException e) {
+                    throw new SystemException(ErrorCode.SYSE0001, e);
                 }
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexUpdaterEvaluatorFactory.java:79]: Add the boolean result:

XDMConstants.setTrue(result);

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexDeleteEvaluatorFactory.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            protected void evaluate(TaggedValuePointable[] args, IPointable result) throws SystemException {
+                IndexUpdater updater = new IndexUpdater(args, result, stringp, bbis, di, sb, abvs, nodeIdProvider,
+                        abvsFileNode, nodep, nodeId);
+                try {
+                    updater.setup();
+                    updater.deleteAllIndexes();
+                } catch (IOException | NoSuchAlgorithmException | JAXBException e) {
+                    throw new SystemException(ErrorCode.SYSE0001, e);
+                }
+
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexDeleteEvaluatorFactory.java:78]: Add the boolean result:

XDMConstants.setTrue(result);

**Comment by prestoncarman** [comment]
After these changes, every thing looks good from my perspective. Lets have Steven do the final review before the merge.

**Comment by prestoncarman** [comment]
Just remembered. These changes to the return type should change the test results.

**Comment by sjaco002** [comment]
This seems fine to me. Please merge with master, squash into a single commit, and make a meaningful commit message. Don't forget to include the issue number in the first line and leave the second line blank

**Pull request closed without merging**